README
NAME
    TryCatch - first class try catch semantics for Perl, without source
    filters.

DESCRIPTION
    This module aims to provide a nicer syntax and method to catch errors in
    Perl, similar to what is found in other languages (such as Java, Python
    or C++). The standard method of using "eval {}; if ($@) {}" is often
    prone to subtle bugs, primarily that its far too easy to stomp on the
    error in error handlers. And also eval/if isn't the nicest idiom.

SYNOPSIS
     use TryCatch;

     sub foo {
       my ($self) = @_;

       try {
         die Some::Class->new(code => 404 ) if $self->not_found;
         return "return value from foo";
       }
       catch (Some::Class $e where { $_->code > 100 } ) {
       }
     }

SYNTAX
    This module aims to give first class exception handling to perl via
    'try' and 'catch' keywords. The basic syntax this module provides is
    "try { # block }" followed by zero or more catch blocks. Each catch
    block has an optional type constraint on it the resembles Perl6's method
    signatures.

    Also worth noting is that the error variable ($@) is localised to the
    try/catch blocks and will not leak outside the scope, or stomp on a
    previous value of $@.

    The simplest case of a catch block is just

     catch { ... }

    where upon the error is available in the standard $@ variable and no
    type checking is performed. The exception can instead be accessed via a
    named lexical variable by providing a simple signature to the catch
    block as follows:

     catch ($err) { ... }

    Type checking of the exception can be performed by specifing a type
    constraint or where clauses in the signature as follows:

     catch (TypeFoo $e) { ... }
     catch (Dict[code => Int, message => Str] $err) { ... }

    As shown in the above example, complex Moose types can be used,
    including MooseX::Types style of type constraints

    In addition to type checking via Moose type constraints, you can also
    use where clauses to only match a certain sub-condition on an error. For
    example, assuming that "HTTPError" is a suitably defined TC:

     catch (HTTPError $e where { $_->code >= 400 && $_->code <= 499 } ) {
       return "4XX error";
     }
     catch (HTTPError $e) {
       return "other http code";
     }

    would return "4XX error" in the case of a 404 error, and "other http
    code" in the case of a 302.

    In the case where multiple catch blocks are present, the first one that
    matches the type constraints (if any) will executed.

BENEFITS
    return. You can put a return in a try block, and it would do the right
    thing - namely return a value from the subroutine you are in, instead of
    just from the eval block.

    Type Checking. This is nothing you couldn't do manually yourself, it
    does it for you using Moose type constraints.

TODO
    *   Decide on "finally" semantics w.r.t return values.

    *   Write some more documentation

    *   Split out the dependancy on Moose

SEE ALSO
    MooseX::Types, Moose::Util::TypeConstraints, Parse::Method::Signatures.

AUTHOR
    Ash Berlin <ash@cpan.org>

THANKS
    Thanks to Matt S Trout and Florian Ragwitz for work on Devel::Declare
    and various B::Hooks modules

    Vincent Pit for Scope::Upper that makes the return from block possible.

    Zefram for providing support and XS guidance.

    Xavier Bergade for the impetus to finally fix this module in 5.12.

LICENSE
    Licensed under the same terms as Perl itself.



Changes
Version history for TryCatch perl module:

1.003002 - 2013/03/21
 - Fix the xml_sax_expat test to check for dependencies correctly. (Who wrote
   that shoddy... oh wait it was me.)

1.003001 - 2013/03/19
  - Fix (the test failure in t/invalid where) syntax errors are reported from
    the wrong location.

1.003000 - 2010/10/13
  - Fix issues with perl 5.12 (finally!)

1.002000 - 2009/09/13
  - Fallback to treating str TCs as class names

1.001001_999 - 2009/09/09
  - Fix for issue with return/unwind when sub is called as a method (vpit)

1.001001_99 - 2009/08/17
  - Fix a few bugs with nested try statements (return and errors not falling
    through as expected)
  - Fix wantarray context of try blocks.

1.001001 - 2009/05/16
  - Fix bug with nested try+die (Reported by Possum)
    As a result of this $@ will not have the value of the exception visible
  - Fix compile issues on 5.8.6 by using ppport.h

1.001000 - 2009/04/26 - 01:01
  - Fix bug where catch block would get run sometimes when it shouldn't
  - Better error reporting
  - Make try 'less reserved' (i.e. so { try => 'foo' } works)
  - Internal refactoring to use Devel::Declare better
  - Switch to using XSLoader from DynaLoader
  - Remove (now redundant) TryCatch::Exception
  - Make things run about 200x faster.

1.000003 - 2009/03/21 - 10:37
  - Makefile.PL tweaks

1.000002 - 2009/03/16 - 21:57
  - Update to more modern Parse::Method::Signatures

1.000001 - 2009/02/14 - 22:16
  - Attempt to improve installability with old CPAN toolchains

1.000000 - 2009/02/07 18:00
  - Initial release


t/lib/NoType.pm
package NoType;

use strict;
use warnings;

#use Moose;
use TryCatch;

BEGIN { 
    package Error;
    sub new { bless {}, $_[0]}
    #use Moose;
}
my $t = Error->new;

sub error {
    die bless {}, 'Error';
}

try {
    error();
}
catch(Error $e) {
    print "Error catched\n";
}

1;


t/lib/NoVarName.pm
package NoVarName;

use strict;
use warnings;

#use Moose;
use TryCatch;

try {
}
catch(Error $) {
    print "Error catched\n";
}

1;



t/lib/TryCatchTest.pm
package TryCatchTest;

sub foo {
  return bar() + baz();
}

sub bar { return 12; }
sub baz { return 30; }

1;


Makefile.PL
use strict;
use warnings;

use FindBin qw/$Bin/;

use inc::Module::Install 0.79;

perl_version '5.008001';
name 'TryCatch';
all_from 'lib/TryCatch.pm';


test_requires 'Test::More' => 0.88;
test_requires 'Test::Exception';

my %config_deps = (
  'ExtUtils::Depends' => 0.302,
);
# These are config and runtime deps
my %run_deps = (
  'B::Hooks::OP::Check' => 0.18,
  'B::Hooks::OP::PPAddr' => 0.03,
);

requires %run_deps;
requires 'Parse::Method::Signatures' => '1.003012';
requires 'B::Hooks::EndOfScope' => 0.12;
requires 'Devel::Declare' => '0.005007';
requires 'Moose';
requires 'MooseX::Types';
requires 'Scope::Upper' => 0.06;
requires 'Variable::Magic' => 0.28;
requires 'Sub::Exporter' => 0.979;
requires 'XSLoader';
requires 'namespace::clean' => '0.20';

resources repository => 'http://github.com/ashb/trycatch/tree/master';

auto_provides;

configure_requires %config_deps, %run_deps;
my $ok = 1;

# Does the toolchain support configure_requires
# If yes, it should be here already.
can_use('ExtUtils::Depends') or do {
  warn "***\n* Your CPAN toolchain is out of date. Please upgrade it\n***\n"; sleep 5;
  requires %config_deps;

};

# Since we can bundle things under inc, do so. That way it should be easier to
# install
unshift @INC, "$Bin/inc/EUDeps";
can_use('ExtUtils::Depends') or do {
  warn "ExtUtils::Depends not installed and can't use version from inc!";
  $ok = 0;
};


unless ($ok) {
  auto_install;
  WriteAll;
  exit 1;
}

my $pkg = ExtUtils::Depends->new('TryCatch',  keys %run_deps);
makemaker_args(
    $pkg->get_makefile_vars,
);

{ no warnings 'once';
if ($Module::Install::AUTHOR) {
  system('pod2text lib/TryCatch.pm > README');

  # Pull in the header files from B::Hooks modules we need.
  # This only works since I know they only use .h files, not XS or typemaps.

  #  Author only - use what ever modules we like!
  require Path::Class;
  require File::Copy::Recursive;

  Path::Class::dir(qw/inc EUDeps/)->rmtree;

  $File::Copy::Recursive::KeepMode = 0;
  my %deps = $pkg->get_deps;
  while (my ($mod, $conf) = each %deps ) {
    my $dest = Path::Class::dir('inc', 'EUDeps', split(/::/, $mod), 'Install', );
    my $src = Path::Class::dir($conf->{instpath});

    File::Copy::Recursive::dircopy("$src", "$dest");
  }
  File::Copy::Recursive::fcopy($INC{'ExtUtils/Depends.pm'}, 'inc/EUDeps/ExtUtils/');
}
}

auto_install;
WriteAll;




eg/bench.pl
#!/usr/bin/perl

use lib 'lib', '../lib';

BEGIN {
    package My::Common;
    *throw_something = $0 =~ /_ok/ ? sub () { 0 } : sub () { 1 };
};

{
    package My::EvalDieObject;
    sub throw {
        my %args = @_;
        die bless {%args}, shift;
    };
};



our %tests = (
    '01_EvalDieScalar' => sub {

        eval {
            die 'Message' if My::Common::throw_something;
        };
        if ($@ =~ /^Message/) {
            1;
        };

    },
    '02_EvalDieObject' => sub {

        eval {
             My::EvalDieObject->throw if My::Common::throw_something;
        };
        if ($@) {
            if (ref $@ && $@->isa('My::EvalDieObject')) {
              my $e = $@;
              1;
            }
        }
    },
);


eval { 
  eval q{use Exception::Base 'Exception::My'};
$tests{ '03_ExceptionEval' } = sub {

        eval {
            Exception::My->throw(message=>'Message') if My::Common::throw_something;
        };
        if ($@) {
            my $e = Exception::Base->catch;
            if ($e->isa('Exception::My') and $e->matches('Message')) {
                1;
            };
        };

    };
$tests{ '04_Exception1Eval' } = sub {

        eval {
            Exception::My->throw(message=>'Message', verbosity=>1) if My::Common::throw_something;
        };
        if ($@) {
            my $e = Exception::Base->catch;
            if ($e->isa('Exception::My') and $e->matches('Message')) {
                1;
            };
        };

    };
};

eval q{
    die 1;
    package My::Error;
    BEGIN {
        eval {
            require Error;
            Error->import(':try');
        };
    };
    Error->VERSION or die;

    $main::tests{'05_Error'} = sub {

        try {
            Error::Simple->throw('Message') if My::Common::throw_something;
        }
        Error->catch(with {
            my $e = $_[0];
            if ($e->text eq 'Message') {
                1;
            }
        });

    };
};

eval {
    package My::ClassThrowable;
    require Class::Throwable;
    Class::Throwable->import;

    $main::tests{'06_ClassThrowable'} = sub {

        eval {
            Class::Throwable->throw('Message') if My::Common::throw_something;
        };
        if ($@) {
            my $e = $@;
            if (ref $e and $e->isa('Class::Throwable') and $e->getMessage eq 'Message') {
                1;
            };
        };

    };
};

eval {
    package My::ExceptionClass;
    require Exception::Class;
    Exception::Class->import('MyException');

    $main::tests{'07_ExceptionClass'} = sub {

        eval {
            MyException->throw(error=>'Message') if My::Common::throw_something;
        };
        my $e;
        if ($e = Exception::Class->caught('MyException') and $e->error eq 'Message') {
            1;
        };

    };
};

eval q{
    package My::ExceptionClassTC;
    BEGIN {
        eval {
            require Exception::Class;
            Exception::Class->import('MyException');
            require Exception::Class::TryCatch;
            Exception::Class::TryCatch->import;
        };
    };
    Exception::Class::TryCatch->VERSION or die;

    $main::tests{'08_ExceptionClassTC'} = sub {

        try eval {
            MyException->throw(error=>'Message') if My::Common::throw_something;
        };
        if (catch my $e) {
            if ($e->isa('MyException') and $e->error eq 'Message') {
                1;
            }
        };

    };
};


{
    package My::TryCatch;
    use TryCatch;

    $main::tests{'09_TryCatch'} = sub {

        try {
            die 'Message this message' if My::Common::throw_something;
        }
        catch ($e) {
            if ($@ =~ /^Message/) {
                1;
            };
        };

    };
};


use Benchmark ':all';

print "Benchmark for ", (My::Common::throw_something ? "FAIL" : "OK"), "\n";
my $result = timethese($ARGV[0] || -1, { %tests });
cmpthese($result);



lib/TryCatch/Exception.pm
package TryCatch::Exception;

use Moose;
use TryCatch qw//;

use MooseX::Types::Moose qw/CodeRef ArrayRef/;

use Scope::Upper qw/unwind want_at :words/;
use namespace::clean -except => 'meta';

has try => (
  is => 'ro',
  isa => CodeRef,
  required => 1
);

has catches => (
  is => 'ro',
  isa => ArrayRef[ArrayRef[CodeRef]],
  default => sub { [] }
);

has ctx => (
  is => 'ro',
  required => 1
);

our $CTX;

sub _run_block {
  my ($self, $code) = (shift,shift);

  my @ret;
  my $wa = want_at $CTX;
  if ($wa) {
    @ret = $code->();
  } elsif (defined $wa) {
    $ret[0] = $code->();
  } else {
    $code->();
  }
  return @ret;
}

sub run {
  my ($self, @args) = @_;
  local $CTX = $CTX;

  my ($package) = caller(1);
  if (defined $package && $package eq __PACKAGE__) {
    # nested: try { try {} }
    die "Internal Error: Nested try without CTX" unless defined $CTX;
  } else {
    $CTX = $self->ctx;
  }

  local $@;
  eval {
    $self->_run_block($self->try, @args);
  };

  # If we get here there was either no explicit return or an error
  return unless $@;
  my $err = $@;

  CATCH: for my $catch ( @{$self->catches} ) {
    my $sub = pop @$catch;
    for my $cond (@$catch) {
      if (ref $cond) {
        local *_;
        $_ = $err;
        next CATCH unless $cond->();
      }
      else {
        my $tc = TryCatch->get_tc($cond);
        next CATCH unless $tc->check($err);
      }
          
    }
    $self->_run_block($sub, @args);
    last;
  }

  return;
}

sub catch {
  my ($self, @conds) = @_;
  push @{$self->catches}, [@conds];
  return $self;

}

__PACKAGE__->meta->make_immutable;

1;


lib/TryCatch/TypeParser.pm
package TryCatch::TypeParser;

# This will eventually belong somewhere else. probably in
# Parse::Method::Signatures
#
# There's probably half a dozen functions in Moose::Util::TypeConstraints I
# should be using instead of what I'm doing here.

use Moose;
use Moose::Meta::TypeConstraint::Union;
use Data::Dumper;
use Carp::Clan qw/^TryCatch/;

use namespace::clean -except => 'meta';

has 'type_registry' => (
  is => 'ro',
  lazy_build => 1
);

has 'package' => (  
  is => 'ro',
  isa => 'ClassName'
);

sub _build_type_registry {
  Moose::Util::TypeConstraints->get_type_constraint_registry;
}

sub str_tc {
  my ($self, $type) = @_;
  
  if (my $code = $self->package->can($type)) {
    my $tc = $code->();
    # TODO: some checking on what $tc is
    return $tc if defined $tc;
  }

  $self->type_registry->find_type_constraint($type) || $type;
}

sub visit {
  my ($self, $tc) = @_;

  unless (ref $tc) {
    return $self->str_tc($tc);
  } elsif (exists $tc->{-or}) {
    return $self->union_tc( $tc->{-or} );
  } elsif ( exists $tc->{-type} ) {
    return $self->param_tc( $tc->{-type}, $tc->{-params} );
  } elsif ( exists $tc->{-str} ) {
    return $tc->{-str};
  } else {
    local $Data::Dumper::Indent = 1;
    croak "TryCatch Internal Error: Cannot deal with tc struct: \n" . Dumper($tc);
  }
}

sub union_tc {
  my ($self, $types) = @_;

  my @types = map { $self->visit($_) } @$types;

  return $types[0] if @types == 1;

  return Moose::Meta::TypeConstraint::Union->new( type_constraints => \@types );
}

sub param_tc {
  my ($self, $type, $params) = @_;
  my @params = map { $self->visit($_) } @$params;

  $type = $self->str_tc($type);

  return $type->parameterize(@params);
}

1;


lib/TryCatch.pm
package TryCatch;

use strict;
use warnings;


use Devel::Declare ();
use B::Hooks::EndOfScope;
use B::Hooks::OP::PPAddr;
use Devel::Declare::Context::Simple;
use Parse::Method::Signatures;
use Moose::Util::TypeConstraints;
use Scope::Upper qw/localize unwind want_at :words/;
use Carp qw/croak/;
use XSLoader;

use base qw/Devel::Declare::Context::Simple/;


our $VERSION = '1.003002';

# Signal to the xs PL_check hooks.
our $NEXT_EVAL_IS_TRY = 0;

# Constants
my ($LOOKAHEAD_TRY, $LOOKAHEAD_CATCH) = (0,1);

XSLoader::load(__PACKAGE__, $VERSION);

use namespace::clean;

use Sub::Exporter -setup => {
  exports => [qw/try/],
  groups => { default => [qw/try/] },
  installer => sub {
    my ($args, $to_export) = @_;
    my $pack = $args->{into};
    my $ctx_class = $args->{parser} || 'TryCatch';

    foreach my $name (@$to_export) {
      if (my $parser = __PACKAGE__->can("_parse_${name}")) {
        Devel::Declare->setup_for(
          $pack,
          { $name => { const => sub { $ctx_class->$parser($pack, @_) } } },
        );
      }
    }
    Sub::Exporter::default_installer(@_);

  }
};


# The actual try call itself. Nothing to do with parsing.
sub try () {
  return;
}

# Where we store all the TCs for catch blocks created at compile time
# Not sure we really want to do this, but we will for now.
our $TC_LIBRARY = {};

sub check_tc {
  my ($class, $tc) = @_;

  my $type = $TC_LIBRARY->{$tc} or die "Unable to find parse TC for '$tc'";

  return $type->check($TryCatch::Error);
}

# From here on out its parsing methods.

# Replace 'try {' with an 'try; { local $@; eval {'
sub _parse_try {
  my ($class,$pack, @args) = @_;

  # Hide from carp - report errors from line of 'try {' in user source.
  local $Carp::Internal{'Devel::Declare'} = 1;

  my $ctx = $class->new->init(@args);

  # Move parse head past 'try ' (space is optional
  $ctx->skip_declarator;
  $ctx->skipspace;

  # Shadow try to be a constant no-op sub. Hopefully
  $ctx->shadow(sub () { } );

  $ctx->inject_if_block(
    $ctx->inject_into_try . $ctx->scope_injector_call( $LOOKAHEAD_TRY ),
    ';'
  ) or croak "block required after try";

  $ctx->debug_linestr('post try');
}

sub scope_injector_call {
  my ($self, $state) = @_;
  return ' BEGIN { ' . ref( $self ) . "->inject_scope($state) }; ";
}


sub inject_scope {
  my ($class, $opts) = @_;

  my $hooks = TryCatch::XS::install_op_checks();

  on_scope_end {
    $class->lookahead_after_block( $opts );

    # TODO: Rethink how i install the hooks. If i uninstall(/disable) them here
    # then they get removed before the LEAVETRY check gets called. Probably
    # switch to a single global set of hooks at look at %^H (?) for lexical
    # goodness.

    #TryCatch::XS::uninstall_op_checks( $hooks );
    #undef $hooks;
  }
}

# Called after the block from try {} or catch {}
#
# Look ahead and determine what action to take based on wether or not we see
# a 'catch' token after the block
sub lookahead_after_block {
  my ($class, $state) = @_;
  my $orig_offset = Devel::Declare::get_linestr_offset();
  my $ctx = $class->new->init( '', $orig_offset );

  my $offset = $ctx->skipspace;
  my $linestr = $ctx->get_linestr;

  my $toke = '';
  my $len = 0;

  # Since we're not being called from a normal D::D callback, we have to
  # find this info manually.
  if ($len = Devel::Declare::toke_scan_word($offset, 1 )) {
    $toke = substr( $linestr, $offset, $len );
    $ctx->{Declarator} = $toke;
  }

  if ($toke eq 'catch') {
    # We don't want the 'catch' token in the output since it messes up the
    # if/else we build up. So don't let control go back to perl just yet.

    $ctx->_parse_catch( $state );

  } else  {
    # No (more) catch blocks, so write the postlude
    my $code;
    if ($state == $LOOKAHEAD_CATCH) {
      $code = $ctx->inject_post_catch;
    }
    else {
      $code = $ctx->inject_when_no_catch;
      $NEXT_EVAL_IS_TRY = 1;
    }

    # Don't try this at home kids
    #
    # Since there was no 'catch' following, move back to the end of the
    # closing brace (where offset was when we started). If we are after the
    # skip space then the 'parse pointer' could be at the start of a POD line,
    # and ";=head1" isn't valid perl ;)
    #
    # This seems to cause problems with nested try so taken out for now
    #
    #TryCatch::XS::set_linestr_offset($orig_offset);
    #$ctx->{Offset} = $orig_offset;

    substr($linestr, $ctx->offset, 0, $code);

    $ctx->set_linestr($linestr);
    $ctx->debug_linestr("finalizer");
  }
}

sub _parse_catch {
  my ($ctx, $state) = @_;

  # Hide these things from carp - this makes C<croak> appear to come from the source line.
  local $Carp::Internal{'TryCatch'} = 1;
  local $Carp::Internal{'Devel::Declare'} = 1;
  local $Carp::Internal{'B::Hooks::EndOfScope::XS'} = 1;
  local $Carp::Internal{'B::Hooks::EndOfScope::PP'} = 1;

  # This isn't a normal DD-callback, so we can strip_name to get rid of 'catch'
  my $offset = $ctx->offset;
  $ctx->strip_name;
  $ctx->skipspace;

  $ctx->debug_linestr('catch');
  my $linestr = $ctx->get_linestr;

  my ($code, $var_code, @conditions) = ("","");

  # optional ()
  if (substr($linestr, $ctx->offset, 1) eq '(') {
    ($var_code, @conditions) = $ctx->parse_proto()
  }

  @conditions = ('1') unless @conditions;

  if ( $state != $LOOKAHEAD_CATCH ) {
    $NEXT_EVAL_IS_TRY = 1;
    $code = $ctx->inject_after_try . "if (";
  }
  else {
    $code = "elsif (";
  }

  $var_code = $ctx->scope_injector_call( $LOOKAHEAD_CATCH ) . $var_code;

  $ctx->inject_if_block(
    $var_code,
    $code . join(' && ', @conditions) . ')'
  ) or croak "block required after catch";

  $ctx->debug_linestr('post catch');

}

sub parse_proto {
  my ($self) = @_;

  my $proto = $self->strip_proto;
  croak "Run-away catch signature"
    unless (length $proto);

  return $self->parse_proto_using_pms($proto);
}

sub _string_to_tc {
  my ($class, $name) = @_;

  my $tc = $class->find_registered_constraint($name);

  return $tc if ref $tc;

  return Moose::Util::TypeConstraints::find_or_create_isa_type_constraint($name)
}

sub parse_proto_using_pms {
  my ($self, $proto) = @_;

  my @conditions;

  my $sig = Parse::Method::Signatures->new(
    input => $proto,
    from_namespace => $self->get_curstash_name,
    type_constraint_callback => \&_string_to_tc,
  );
  my $errctx = $sig->ppi;
  my $param = $sig->param;

  $sig->error( $errctx, "Parameter expected")
    unless $param;

  my $left = $sig->remaining_input;

  my $var_code = '';

  if (my $var_name = $param->can('variable_name') ) {

    my $name = $param->$var_name();
    $var_code = "my $name = \$TryCatch::Error;";
  }

  # (TC $var)
  if ($param->has_type_constraints) {
    my $tc = $param->meta_type_constraint;
    $TryCatch::TC_LIBRARY->{"$tc"} = $tc;
    push @conditions, "TryCatch->check_tc('$tc')";
  }

  # ($var where { $_ } )
  if ($param->has_constraints) {
    foreach my $con (@{$param->constraints}) {
      $con =~ s/^{|}$//g;
      push @conditions, "do {local \$_ = \$TryCatch::Error; $con }";
    }
  }

  return $var_code, @conditions;
}


#######################################################################
# Injected snippets

sub inject_into_try {
  # try { ...
  # ->
  # try; { local $@; eval { ...

  'local $@; eval {'
}

sub inject_after_try {
  # This semicolon is for the end of the eval
  return ';$TryCatch::Error = $@; } if ($TryCatch::Error) { ';
}

sub inject_when_no_catch {
  # This undef is to ensure that there is the eval{}; is called in void context
  # i.e that its not the last op in a subroutine
  return "};undef;";
}

sub inject_post_catch {
  # We do it like this so that PROPGATE gets called, in case anyone is using it
  return 'else { $@ = $TryCatch::Error; die } };undef;';
}

#######################################################################

require Devel::PartialDump if $ENV{TRYCATCH_DEBUG};

*debug_linestr = !( ($ENV{TRYCATCH_DEBUG} || 0) & 1)
               ? sub {}
               : sub {
  my ($ctx, $message) = @_;

  local $Carp::Internal{'TryCatch'} = 1;
  local $Carp::Internal{'TryCatch::Basic'} = 1;
  local $Carp::Internal{'Devel::Declare'} = 1;
  local $Carp::Internal{'B::Hooks::EndOfScope:XS'} = 1;
  local $Carp::Internal{'B::Hooks::EndOfScope:PP'} = 1;
  local $Carp::Internal{'Devel::PartialDump'} = 1;
  Carp::cluck($message) if $message;

  warn   "  Substr: ", Devel::PartialDump::dump(substr($ctx->get_linestr, $ctx->offset)),
       "\n  Whole:  ", Devel::PartialDump::dump($ctx->get_linestr), "\n\n";
};


1;

__END__

=head1 NAME

TryCatch - first class try catch semantics for Perl, without source filters.

=head1 DESCRIPTION

This module aims to provide a nicer syntax and method to catch errors in Perl,
similar to what is found in other languages (such as Java, Python or C++).  The
standard method of using C<< eval {}; if ($@) {} >> is often prone to subtle
bugs, primarily that its far too easy to stomp on the error in error handlers.
And also eval/if isn't the nicest idiom.

=head1 SYNOPSIS

 use TryCatch;

 sub foo {
   my ($self) = @_;

   try {
     die Some::Class->new(code => 404 ) if $self->not_found;
     return "return value from foo";
   }
   catch (Some::Class $e where { $_->code > 100 } ) {
   }
 }

=head1 SYNTAX

This module aims to give first class exception handling to perl via 'try' and
'catch' keywords. The basic syntax this module provides is C<try { # block }>
followed by zero or more catch blocks. Each catch block has an optional type
constraint on it the resembles Perl6's method signatures.

Also worth noting is that the error variable (C<$@>) is localised to the
try/catch blocks and will not leak outside the scope, or stomp on a previous
value of C<$@>.

The simplest case of a catch block is just

 catch { ... }

where upon the error is available in the standard C<$@> variable and no type
checking is performed. The exception can instead be accessed via a named
lexical variable by providing a simple signature to the catch block as follows:

 catch ($err) { ... }

Type checking of the exception can be performed by specifying a type constraint
or where clauses in the signature as follows:

 catch (TypeFoo $e) { ... }
 catch (Dict[code => Int, message => Str] $err) { ... }

As shown in the above example, complex Moose types can be used, including
L<MooseX::Types> style of type constraints

In addition to type checking via Moose type constraints, you can also use where
clauses to only match a certain sub-condition on an error. For example,
assuming that C<HTTPError> is a suitably defined TC:

 catch (HTTPError $e where { $_->code >= 400 && $_->code <= 499 } ) {
   return "4XX error";
 }
 catch (HTTPError $e) {
   return "other http code";
 }

would return "4XX error" in the case of a 404 error, and "other http code" in
the case of a 302.

In the case where multiple catch blocks are present, the first one that matches
the type constraints (if any) will executed.

=head1 BENEFITS

B<return>. You can put a return in a try block, and it would do the right thing
- namely return a value from the subroutine you are in, instead of just from
the eval block.

B<Type Checking>. This is nothing you couldn't do manually yourself, it does it
for you using Moose type constraints.

=head1 TODO

=over

=item *

Decide on C<finally> semantics w.r.t return values.

=item *

Write some more documentation

=item *

Split out the dependency on Moose

=back

=head1 SEE ALSO

L<MooseX::Types>, L<Moose::Util::TypeConstraints>, L<Parse::Method::Signatures>.

=head1 AUTHOR

Ash Berlin <ash@cpan.org>

=head1 THANKS

Thanks to Matt S Trout and Florian Ragwitz for work on L<Devel::Declare> and
various B::Hooks modules

Vincent Pit for L<Scope::Upper> that makes the return from block possible.

Zefram for providing support and XS guidance.

Xavier Bergade for the impetus to finally fix this module in 5.12.

=head1 LICENSE

Licensed under the same terms as Perl itself.

